MdePkg/Library/UefiUsbLib/UsbDxeLib.c

```diff
@@ -650,3 +654,512 @@ UsbClearEndpointHalt (
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  DescLength = ((UINT8 *)Data)[0];
+
+  if ((Buffer == NULL) || (DescLength > *BufferSize)) {
+    *BufferSize = DescLength;
+    return EFI_BUFFER_TOO_SMALL;
+  }
+
+  CopyMem (Buffer, Data, DescLength);
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Retrieve class specific endpoint descriptor.
+
+  @param[in]  This              A pointer to the EFI_USB_IO_PROTOCOL instance.
+  @param[in]  Setting           Interface alternate setting.
+  @param[in]  Index             Zero-based index of the non-zero endpoint.
+  @param[in][out]  BufferSize   On input, the size in bytes of the return Descriptor buffer.
+                                On output the size of data returned in Descriptor.
+  @param[out]  Descriptor       The buffer to return the contents of the class specific endpoint descriptor. May
+                                be NULL with a zero BufferSize in order to determine the size buffer needed.
+
+  @retval EFI_SUCCESS           Output parameters were updated successfully.
+  @retval EFI_INVALID_PARAMETER BufferSize is NULL.
+                                Buffer is NULL and *BufferSize is not zero.
+  @retval EFI_UNSUPPORTED       Setting is greater than the number of alternate settings in this interface.
+  @retval EFI_NOT_FOUND         Index is greater than the number of endpoints in this interface.
+                                Endpoint does not have class specific endpoint descriptor.
+  @retval EFI_BUFFER_TOO_SMALL  The BufferSize is too small for the result. BufferSize has been updated with the size
+                                needed to complete the request.
+  @retval EFI_DEVICE_ERROR      Error reading device data.
+
+**/
+EFI_STATUS
+EFIAPI
+UsbGetCsEndpointDescriptor (
+  IN  EFI_USB_IO_PROTOCOL  *This,
+  IN  UINTN                Setting,
+  IN  UINTN                Index,
+  IN OUT UINTN             *BufferSize,
+  OUT VOID                 *Buffer
+  )
+{
+  EFI_STATUS  Status;
+  VOID        *Data;
+  UINT8       DescLength;
+
+  if ((BufferSize == NULL) || ((Buffer == NULL) && (*BufferSize != 0))) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Status = InitUsbConfigDescriptorData (This);
+  if (EFI_ERROR (Status)) {
+    return EFI_DEVICE_ERROR;
+  }
+
+  Status = FindUsbDescriptor (This, USB_DESC_TYPE_CS_ENDPOINT, Setting, Index, &Data);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  DescLength = ((UINT8 *)Data)[0];
+
+  if ((Buffer == NULL) || (DescLength > *BufferSize)) {
+    *BufferSize = DescLength;
+    return EFI_BUFFER_TOO_SMALL;
+  }
+
+  CopyMem (Buffer, Data, DescLength);
+
+  return EFI_SUCCESS;
+}
+
+/**
+  The constructor function initializes global pointer to the configuration data.
+
+  @param  ImageHandle   The firmware allocated handle for the EFI image.
+  @param  SystemTable   A pointer to the EFI System Table.
+
+  @retval EFI_SUCCESS   The constructor always returns EFI_SUCCESS.
+
+**/
+EFI_STATUS
+EFIAPI
+UefiUsbLibConstructor (

```

Comment:
This constructor can be removed if the global variable is pre-initialized to NULL in its declaration.

Reply:
done